File: neet_code_all_cpp/math_and_geometry/ugly_number.cpp
#include <iostream>

class Solution {
    public:
        // AC
        bool isUgly(int n) {
            if(n < 0) {
                return false;
            }

            while(n != 0 && n % 2 == 0) {
                n /= 2;
            }
            while(n != 0 && n % 3 == 0) {
                n /= 3;
            }
            while(n != 0 && n % 5 == 0) {
                n /= 5;
            }

            return n == 1;
        }
};

// 模範解答
class SolutionAns {
    public:
        bool isUgly(int n) {
            if(n <= 0) {
                return false;
            }

            for(auto p: {2, 3, 5}) {
                while(n % p == 0) {
                    n /= p;
                }
            }

            return n == 1;
        }
};

int main(void) {
    int case_1 = 6;
    int case_2 = 1;
    int case_3 = 14;

    Solution s_1;

    std::cout << s_1.isUgly(case_1) << std::endl;
    std::cout << s_1.isUgly(case_2) << std::endl;
    std::cout << s_1.isUgly(case_3) << std::endl;

    SolutionAns s_ans;

    std::cout << s_ans.isUgly(case_1) << std::endl;
    std::cout << s_ans.isUgly(case_2) << std::endl;
    std::cout << s_ans.isUgly(case_3) << std::endl;
}

File: neet_code_all/src/math_and_geometry/ugly_number.rs
struct Solution;
impl Solution {
    // AC
    pub fn is_ugly(n: i32) -> bool {
        if n < 0 {
            return false;
        }
        let mut n = n;

        while n != 0 && n % 2 == 0 {
            n /= 2;
        }
        while n != 0 && n % 3 == 0 {
            n /= 3;
        }
        while n != 0 && n % 5 == 0 {
            n /= 5;
        }

        n == 1
    }
}

// 模範解答
struct SolutionAns;
impl SolutionAns {
    pub fn is_ugly(n: i32) -> bool {
        if n < 1 {
            return false;
        }
        let mut n = n;

        for p in [2, 3, 5] {
            while n % p == 0 {
                n /= p;
            }
        }

        n == 1
    }
}

fn main() {
    let case_1 = 6;
    let case_2 = 1;
    let case_3 = 14;

    println!("case_1: {}", Solution::is_ugly(case_1));
    println!("case_2: {}", Solution::is_ugly(case_2));
    println!("case_3: {}", Solution::is_ugly(case_3));

    println!("case_1: {}", SolutionAns::is_ugly(case_1));
    println!("case_2: {}", SolutionAns::is_ugly(case_2));
    println!("case_3: {}", SolutionAns::is_ugly(case_3));
}
